base/applications/cmdutils/taskkill/taskkill.c

```diff
@@ -249,57 +356,90 @@ static int send_close_messages(void)
-        {
-            DWORD index;
-            BOOL found_process = FALSE;
 
-            for (index = 0; index < pid_list_size; index++)
+        HeapFree(GetProcessHeap(), 0, pkill_list);
+    }
+
+    HeapFree(GetProcessHeap(), 0, pid_list);
+    return status_code;
+}
+
+static void terminate_process_tree(DWORD ppid)
+{
+    FILETIME parent_creation_time = { 0 };
+    HANDLE h = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
+    PROCESSENTRY32W pe = { 0 };
+    pe.dwSize = sizeof(PROCESSENTRY32W);
+
+    if (!get_pid_creation_time(ppid, &parent_creation_time) || !h)
+    {
+        CloseHandle(h);
+        return;
+    }
+
+    if (Process32FirstW(h, &pe))
+    {
+        do {
+            FILETIME child_creation_time = { 0 };
+
+            if (!get_pid_creation_time(pe.th32ProcessID, &child_creation_time))
             {
-                WCHAR process_name[MAX_PATH];
+                continue;
+            }
 
-                if (get_process_name_from_pid(pid_list[index], process_name, MAX_PATH) &&
-                    !strcmpiW(process_name, task_list[i]))
+            // Compare creation time to avoid reuse PID, thanks to @ThFabba
+            if (pe.th32ParentProcessID == ppid &&
+                CompareFileTime(&parent_creation_time, &child_creation_time) < 0)
+            {
+                HANDLE process = INVALID_HANDLE_VALUE;
+
+                // Use recursion to browse all child processes
+                terminate_process_tree(pe.th32ProcessID);
+                process = OpenProcess(PROCESS_TERMINATE, FALSE, pe.th32ProcessID);
+                if (!process)
                 {
-                    struct pid_close_info info = { pid_list[index] };
-
-                    found_process = TRUE;
-                    if (pid_list[index] == self_pid)
-                    {
-                        taskkill_message(STRING_SELF_TERMINATION);
-                        status_code = 1;
-                        continue;
-                    }
-
-                    EnumWindows(pid_enum_proc, (LPARAM)&info);
-                    taskkill_message_printfW(STRING_CLOSE_PROC_SRCH, process_name, pid_list[index]);
+                    continue;
                 }
-            }
 
-            if (!found_process)
-            {
-                taskkill_message_printfW(STRING_SEARCH_FAILED, task_list[i]);
-                status_code = 128;
+                if (!TerminateProcess(process, 0))
+                {
+                    taskkill_message_printfW(STRING_TERM_CHILD_FAILED, pe.th32ProcessID, ppid);
+                    CloseHandle(process);
+                    continue;
+                }
+
+                taskkill_message_printfW(STRING_TERM_CHILD, pe.th32ProcessID, ppid);
+                CloseHandle(process);
             }
-        }
+        } while (Process32NextW(h, &pe));
     }
 
-    HeapFree(GetProcessHeap(), 0, pid_list);
-    return status_code;
+    CloseHandle(h);
 }
 
 static int terminate_processes(void)
 {
```

Comment:
_**Optional comment for the time being**_
If you look closely, this function is extremely similar to `send_close_messages()`, except when it calls its corresponding helpers. Therefore I think it should be possible to merge both functions together into a single one that would do the common job currently done by these two functions, namely browsing through the list of tasks... and then make that new single function call different helper callback functions depending on whether one wants to just send close messages, or to actually terminate processes.

Reply:
Yea, I thought about it. But I saw they use different message format to print so I decided keep them separate. I'll merge them and use callback functions.

Reply:
Sorry, I can't merge them. Because the message formats used by `send_close_messages()`,  `terminate_processes()` , `send_close_messages_tree(DWORD ppid)` and `terminate_process_tree(DWORD ppid)` are different so I don't know the way to handle.

Reply:
OK so the functionality of these functions have been already merged in master (with the commit of @kernelbin), so you don't need to worry about that.